### PR TITLE
Fix: Persisting Active Style on List and DataTable

### PR DESCRIPTION
**Pull request by jlaluces123**
Addresses #5266 : When selecting a row from a list or table, the active styling on the item remains when a user opens a modal / layer detail, but when closing the modal, the active styling disappears, preventing users from seeing where they last left off.

##### List Component:
- Persists the active styling during an `onBlur` and `onMouseOut` 
- Updates to the `onClickItem` story to match the scenario found in issue #5266
- `lastActive` pointer to keep track of the latest active item / row selected to persist active styling

##### DataTable Component:
- Updates to the `clickable` story to match scenario found in issue #5266
- In Progress: persisting the active style during `onBlur` and `onMouseLeave`

##### List Component:
1. In `components/List/List.js` check out the `clickProps` to see how the active styling is applied / persisting depending on the events
2. `yarn storybook`
3. Navigate to `List's onClickItem` story
4. Click on any list item, close the modal and watch how the active styling remains until you hover other list items

##### DataTable Component
1. In `components/DataTable/Body.js` check out the `clickProps (onMouseEnter, onFocus, [...])` to see how the active styling is applied / persisting depending on the events
2. `yarn storybook`
3. Navigate to `DataTable's Clickable` story
4. Work in Progress: click on any row, close the modal and watch how the active styling remains until you hover other list items

#### What testing has been done on this PR?
Manual testing via the onClickItem story.

#### How should this be manually tested?
1. `yarn storybook`
2. Navigate to `List's onClickItem` story
3. Click on any list item, close the modal and watch how the active styling remains until you hover other list items
4. Repeat Steps 2 and 3 for `DataTable's Clickable` story

#### What are the relevant issues?
- #5266

#### Do the grommet docs need to be updated?
No.

#### Should this PR be mentioned in the release notes?
No.

#### Is this change backwards compatible or is it a breaking change?
This change will be backwards compatible.

### CODE REVIEW

File: src/js/components/DataTable/DataTable.js [hunk 1]
```diff
@@ -164,6 +164,7 @@ const DataTable = ({
   const footerRef = useRef();
   const [headerHeight, setHeaderHeight] = useState();
   const [footerHeight, setFooterHeight] = useState();
+  // const [lastActive, setLastActive] = useState();
```
Comment (ShimiSun) [inline src/js/components/DataTable/DataTable.js:167]: we try not to have code comments on our code base, to avoid clutter.

**Review by ShimiSun — COMMENTED** [review]
Some of the auto spacing on this PR isn't aligned with our original code base guidelines, it will be helpful if the changed code won't show spacing and braces changes if not necessary.

**Comment by ericsoderberghp** [comment]
I'm noticing a couple of things about the behavior in Storybook that don't look quite right:

List

I can tab to the List and see the focus outline
I can use the arrow keys to move the active item indicator up and down
I can use the enter key to show the layer
I can use the esc key to close the layer and the focus and active indicator look correct
But, if I now tab to remove focus from the List, the active list item indicator remains. I think the active item indicator shouldn't be shown if the List doesn't have focus.

DataTable

If I tab to the DataTable, I don't see the focus indicator, but I should.
I can use the arrow keys to move the active row indicator up and down.
I can use the enter key to show the modal
I can use the esc key to close the modal and the active indicator looks correct, but I still don't see the focus indicator.
If I tab away from the DataTable, the active row indicator remains. I think the active row indicator shouldn't be shown if the DataTable doesn't have focus.

File: src/js/components/DataTable/Body.js [hunk 2]
```diff
@@ -44,6 +48,7 @@ const Row = memo(
       <StyledDataTableRow
         ref={rowRef}
         size={size}
+        tabIndex={onClickRow ? -1 : undefined}
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:51]: Why is this needed? I wasn't expecting us to be giving focus to the row.

File: src/js/components/DataTable/Body.js [hunk 3]
```diff
@@ -218,10 +250,15 @@ const Body = forwardRef(
...
                 <Row
                   key={index}
+                  tableRef={ref}
+                  lastActive={lastActive}
+                  setLastActive={setLastActive}
                   rowRef={rowRef}
                   primaryValue={primaryValue}
                   isSelected={isSelected}
                   isRowExpanded={isRowExpanded}
+                  itemFocus={itemFocus}
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:260]: I didn't see where this was used by Row. Do we need to pass it here?

File: src/js/components/DataTable/StyledDataTable.js [hunk 4]
```diff
@@ -103,12 +108,18 @@ const StyledDataTableBody = styled(TableBody)`
     overflow: auto;
   `}
 
-  &:focus {
-    ${focusStyle({ skipSvgChildren: true, forceOutline: true })}
-  }
+  ${props =>
+    props.itemFocus &&
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/StyledDataTable.js:112]: I don't think the StyledDataTableBody ever received `itemFocus` as a property.

File: src/js/components/DataTable/StyledDataTable.js [hunk 5]
```diff
@@ -103,12 +108,18 @@ const StyledDataTableBody = styled(TableBody)`
     overflow: auto;
   `}
 
-  &:focus {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/StyledDataTable.js:106]: Why did we not keep focus at the table body?

File: src/js/components/DataTable/Body.js [hunk 6]
```diff
@@ -54,13 +59,37 @@ const Row = memo(
...
                 onClickRow(adjustedEvent);
+                // putting focus on tbody for WCAG guideline
+                tableRef.current.focus();
+              }
+            : undefined
+        }
+        onMouseEnter={
+          onClickRow
+            ? () => {
+                setLastActive(undefined);
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:70]: I'm thinking we might want to simplify this and put more logic in the Body than the Row, the Body is the one keeping track of the active state. For example, the Body can manage `lastActive` as it sees `active` change.

File: src/js/components/DataTable/StyledDataTable.js [hunk 7]
```diff
@@ -83,6 +83,11 @@ const StyledDataTableRow = styled(TableRow)`
     `
     cursor: pointer;
   `}
+
+  &:focus {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/StyledDataTable.js:87]: Just wondering why we are changing the pattern and putting focus on the row instead of the body.

File: src/js/components/List/List.js [hunk 8]
```diff
@@ -232,10 +232,8 @@ const List = React.forwardRef(
             (onClickItem || onOrder) && data && data.length
               ? () => {
                   const min = onOrder ? 1 : 0;
-                  const max = onOrder ? (data.length * 2) - 2 : data.length - 1;
-                  setActive(
-                    active >= min ? Math.min(active + 1, max) : min,
-                  );
+                  const max = onOrder ? data.length * 2 - 2 : data.length - 1;
```
Comment (ericsoderberghp) [inline src/js/components/List/List.js:235]: Why are all of these syntax changes being included? They don't seem related to this active state change.

File: src/js/components/List/List.js [hunk 9]
```diff
@@ -22,11 +22,13 @@ const StyledList = styled.ul`
   padding: 0;
   ${genericStyles}
 
+  // Customizes to make table have a focus border color of green
```
Comment (ericsoderberghp) [inline src/js/components/List/List.js:25]:
```suggestion
  // Customizes to make list have a focus border color of green
```

File: src/js/components/List/List.js [hunk 10]
```diff
@@ -220,22 +222,23 @@ const List = React.forwardRef(
                 }
               : undefined
           }
+          onTab={() => setActive(undefined)}
```
Comment (ericsoderberghp) [inline src/js/components/List/List.js:225]: Why do we need the `onTab` handler? I would expect `onBlur` to be sufficient.

File: src/js/components/List/List.js [hunk 11]
```diff
@@ -357,13 +366,19 @@ const List = React.forwardRef(
...
                       listRef.current.focus();
                     },
-                    onMouseOver: () => setActive(index),
-                    onMouseOut: () => setActive(undefined),
+                    onMouseOver: () => {
+                      setLastActive(undefined);
+                      setActive(index);
+                    },
+                    onMouseOut: () => {
+                      setActive(lastActive);
```
Comment (ericsoderberghp) [inline src/js/components/List/List.js:374]: This doesn't seem right. I would think we would leave `lastActive` alone, to be used in the next `onFocus`, but we wouldn't want to `setActive` when the mouse goes away.

File: src/js/components/List/List.js [hunk 12]
```diff
@@ -357,13 +366,19 @@ const List = React.forwardRef(
...
+                    },
                     onFocus: () => {
                       setActive(index);
                       setItemFocus(true);
                     },
                     onBlur: () => {
+                      setLastActive(active);
```
Comment (ericsoderberghp) [inline src/js/components/List/List.js:381]: I would have expected `onFocus` to set both `active` and `lastActive` and `onBlur` to just clear `active`. Why are we setting last active at `onBlur` time?

**Review by ericsoderberghp — COMMENTED** [review]
I've added some some comments to List for now. It looks like there are still unresolved comments in DataTable from prior reviews.

File: src/js/components/List/List.js (hunk 12 above)
Comment (jlaluces123) [reply src/js/components/List/List.js:381]: I agree with this idea, moving the `lastActive` to `onFocus` makes more sense here

File: src/js/components/List/List.js (hunk 10 above)
Comment (jlaluces123) [reply src/js/components/List/List.js:225]: From what I gathered, since `onBlur` is only part of the `clickProps` events, `onBlur` and `onFocus` never get triggered when doing keyboard navigation which is why I decided to use the `onTab` prop available for Keyboard as a way to remove the active indicator upon losing focus of the List or DataTable.

File: src/js/components/List/List.js (hunk 11 above)
Comment (jlaluces123) [reply src/js/components/List/List.js:374]: As soon as the modal / Layer opens, `onMouseOut` is triggered and upon closing, I chose to `setActive`  to what `lastActive` had stored as the way to keep an item active when a modal / Layer closes.

File: src/js/components/DataTable/Body.js (hunk 2 above)
Comment (jlaluces123) [reply src/js/components/DataTable/Body.js]: I played around with this and the tabIndex was needed in order to use the `onFocus` and `onBlur` properties but since we don't want to put focus on the row, I set it to `-1` to counteract that.

File: src/js/components/DataTable/Body.js (hunk 3 above)
Comment (jlaluces123) [reply src/js/components/DataTable/Body.js]: Got rid of this here after trimming the unused code as it was incorrectly placed, this was supposed to be put on `StyledDataTableBody`

File: src/js/components/DataTable/StyledDataTable.js (hunk 4 above)
Comment (jlaluces123) [reply src/js/components/DataTable/StyledDataTable.js:114]: Resolved now, accidentally placed `itemFocus` on the Row component instead of the Table Body

File: src/js/components/DataTable/StyledDataTable.js (hunk 5 above)
Comment (jlaluces123) [reply src/js/components/DataTable/StyledDataTable.js:106]: This is back in place now but utilizing the same approach as List.js where it'll look like:

```js
&:focus {
  ${props => props.tabIndex >= 0 && focusStyle ...}
}
```

File: src/js/components/DataTable/StyledDataTable.js (hunk 7 above)
Comment (jlaluces123) [reply src/js/components/DataTable/StyledDataTable.js]: As part of the trimming down I've done, this will be removed as well. I believe where we check for `props.active` that mimics the same functionality.

**Comment by ShimiSun** [comment]
Closing as this PR has an updated version #5371
